### PR TITLE
introduce nondet_symbol_exprt

### PR DESCRIPTION
**Pull request by kroening**
This introduces a class for an expression type that is rare but has been around for some time.

### CODE REVIEW

**Comment by smowton** [comment]
How does it differ from side_effect_expr_nondett assigned to a regular symbol?

File: src/solvers/smt2/smt2_conv.cpp [hunk 1]
```diff
@@ -853,14 +853,13 @@ void smt2_convt::convert_expr(const exprt &expr)
   if(expr.id()==ID_symbol)
   {
     irep_idt id=to_symbol_expr(expr).get_identifier();
-    assert(id!=irep_idt());
-
+    DATA_INVARIANT(id!=irep_idt(), "symbol must have identifier");
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_conv.cpp]: Use `!id.empty()`? (And actually make the same change in smt1.)

File: src/solvers/smt2/smt2_conv.cpp [hunk 2]
```diff
@@ -853,14 +853,13 @@ void smt2_convt::convert_expr(const exprt &expr)
...
     out << '|' << convert_identifier(id) << '|';
   }
   else if(expr.id()==ID_nondet_symbol)
   {
-    irep_idt id=expr.get(ID_identifier);
-    assert(id!="");
+    irep_idt id=to_nondet_symbol_expr(expr).get_identifier();
+    DATA_INVARIANT(id!=irep_idt(), "symbol must have identifier");
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_conv.cpp]: As above: `!id.empty()`

**Review by tautschnig — COMMENTED** [review]
It would be nice if the commit message were more verbose, answering @smowton's question by explaining that one is code (with side effects: repeated execution will yield new symbols) while this one is a non-deterministic constant.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**